gk/fib.c

```diff
@@ -1254,34 +1269,45 @@ init_grantor_fib_locked(struct ip_prefix *ip_prefix,
 		num_cache_holds++;
 	}
 
-	fib_id = get_empty_fib_id(ip_prefix->addr.proto, gk_conf);
-	if (fib_id < 0)
-		goto put_ether_cache;
+	if (!prefix_exists) {
+		fib_id = get_empty_fib_id(ip_prefix->addr.proto, gk_conf);
+		if (fib_id < 0)
+			goto put_ether_cache;
+	}
 
 	if (ip_prefix->addr.proto == RTE_ETHER_TYPE_IPV4)
 		gt_fib = &ltbl->fib_tbl[fib_id];
 	else
 		gt_fib = &ltbl->fib_tbl6[fib_id];
 
-	gt_fib->action = GK_FWD_GRANTOR;
-	gt_fib->u.grantor.set = rte_malloc("gk_fib.grantor.set",
-		sizeof(*(gt_fib->u.grantor.set)) +
-			num_addrs * sizeof(*(gt_fib->u.grantor.set->entries)),
-		0);
-	if (gt_fib->u.grantor.set == NULL) {
+	new_set = rte_malloc("gk_fib.grantor.set",
+		sizeof(*new_set) + num_addrs * sizeof(*(new_set->entries)), 0);
+	if (new_set == NULL) {
 		GK_LOG(ERR, "Could not allocate set of Grantor entries\n");
 		goto init_fib;
 	}
-	gt_fib->u.grantor.set->proto = ip_prefix->addr.proto;
-	gt_fib->u.grantor.set->num_entries = num_addrs;
+	new_set->proto = ip_prefix->addr.proto;
+	new_set->num_entries = num_addrs;
 	for (i = 0; i < num_addrs; i++) {
-		gt_fib->u.grantor.set->entries[i].gt_addr = gt_addrs[i];
-		gt_fib->u.grantor.set->entries[i].eth_cache = eth_caches[i];
+		new_set->entries[i].gt_addr = gt_addrs[i];
+		new_set->entries[i].eth_cache = eth_caches[i];
 	}
 
-	ret = lpm_add_route(&ip_prefix->addr, ip_prefix->len, fib_id, ltbl);
-	if (ret < 0)
-		goto free_set;
+	if (prefix_exists) {
+		/* Replace old set of Grantors in existing entry. */
+		struct grantor_set *old_set = gt_fib->u.grantor.set;
+		gt_fib->u.grantor.set = new_set;
+		synchronize_gk_instances(gt_fib, gk_conf);
```

Comment:
`gk_synchronize()`, the function that is going to be triggered at each GK block, will flush the flow table because it only checks `fib->action`, so it doesn't know that only the grantor list is being updated in `gt_fib`.